File: app/src/main/java/com/allybros/superego/activity/SplashActivity.java
package com.allybros.superego.activity;


import static com.allybros.superego.unit.ConstantValues.IS_SHOWN;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.util.Log;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.allybros.superego.R;
import com.allybros.superego.api.GetAlertsTask;
import com.allybros.superego.api.LoadProfileTask;
import com.allybros.superego.api.mapper.UserMapper;
import com.allybros.superego.unit.Alert;
import com.allybros.superego.unit.ErrorCodes;
import com.allybros.superego.unit.User;
import com.allybros.superego.util.ClientContextUtil;
import com.allybros.superego.util.SessionManager;
import com.google.gson.Gson;


@SuppressLint("CustomSplashScreen")
public class SplashActivity extends AppCompatActivity {
    private static boolean alertsShown = false; // Do not show alerts multiple times to user
    private ActivityResultLauncher<Intent> alertActivityLauncher;
    private ActivityResultLauncher<Intent> guideActivityLauncher;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_splash);

        registerAlertActivityLauncher();
        registerGuideActivityLauncher();

        if (!checkNetworkConnection()) {
            // No internet connection
            showNetworkErrorDialog();
        } else if (!isGuideShown()) {
            // Show guide activity
            startGuideActivity();
        } else {
            // Ask for alerts
            onReadyForAlerts();
        }
    }

    /**
     * Performs check on internet connection
     * @return false if no network connectivity
     */
    private boolean checkNetworkConnection() {
        ConnectivityManager cm = (ConnectivityManager) getApplicationContext().getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
        return activeNetwork != null && activeNetwork.isConnectedOrConnecting();
    }

    /**
     * Shows network error dialog
     */
    private void showNetworkErrorDialog() {
        new AlertDialog.Builder(this, R.style.SegoAlertDialog)
                .setTitle("insightof.me")
                .setMessage(R.string.error_no_connection)
                .setPositiveButton(getString(R.string.action_ok), (dialog, which) -> {
                    SplashActivity.this.finish();
                    System.exit(0);
                })
                .show();
    }

    /**
     * Initializes alert activity launcher
     */
    private void registerAlertActivityLauncher() {
        // Set ready for login as the callback
        this.alertActivityLauncher = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                o -> onReadyForLogin()
        );
    }

    /**
     * Initializes guide activity launcher
     */
    private void registerGuideActivityLauncher() {
        // Set ready for alerts as the callback
        this.guideActivityLauncher = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                o -> onReadyForAlerts()
        );
    }


    /**
     * Send requests for alerts
     */
    private void onReadyForAlerts() {
        if (alertsShown) {
            Log.i("ReadyForAlerts", "Alerts already shown to the user");
            onReadyForLogin();

        } else {
            Log.i("ReadyForAlerts", "Alerts not shown before");
            executeGetAlertsTask();
        }
    }

    /**
     * Send requests and launch the applications
     */
    private void onReadyForLogin() {
        Log.i(getClass().getSimpleName(), "Ready for starting application");
        // Read session data
        SessionManager.getInstance().readInfo(getApplicationContext());
        if (!SessionManager.getInstance().getSessionToken().isEmpty()) {
            // User signed in before
            Log.i("ReadyForStart", "Session token found, starting load profile task");
            executeLoadProfileTask(SessionManager.getInstance().getSessionToken());
        } else {
            // No session, go to login
            Log.i("ReadyForStart", "No session data found, redirecting to login");
            startLoginActivity();
        }
    }

    private void executeLoadProfileTask(String sessionToken) {
        LoadProfileTask loadProfileTask = new LoadProfileTask(sessionToken);
        loadProfileTask.setOnResponseListener(response -> {
            // Map to session user
            if (response.getStatus() == ErrorCodes.SUCCESS) {
                // Successfully response
                User sessionUser = UserMapper.fromProfileResponse(response);
                SessionManager.getInstance().setUser(sessionUser);
                // Start user page activity
                startUserPageActivity();
            } else {
                Log.w("Splash", "Can not load profile, routing to login page");
                startLoginActivity();
            }

        });
        loadProfileTask.execute(this);
    }

    private void startGuideActivity() {
        Intent intent = new Intent(getApplicationContext(), GuideActivity.class);
        this.guideActivityLauncher.launch(intent);
    }

    private boolean isGuideShown() {
        SharedPreferences pref = getSharedPreferences(IS_SHOWN, MODE_PRIVATE);
        return pref.getBoolean("isShown", false);
    }

    /**
     * Checks if both task are completed. If both of them completed, starts UserPageActivity
     */
    private synchronized void startUserPageActivity(){
        // Both task are completed
        Intent i = new Intent(SplashActivity.this, UserPageActivity.class);
        i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(i);
        finish();

    }

    /**
     * Starts login activity and finished this activity
     */
    private void startLoginActivity(){
        Intent i = new Intent(SplashActivity.this, LoginActivity.class);
        i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(i);
        finish();
    }

    /**
     * Starts get alerts task
     */
    private void executeGetAlertsTask() {
        String versionName = ClientContextUtil.getVersionName(this);
        GetAlertsTask alertsApiTask = new GetAlertsTask(versionName);
        alertsApiTask.setOnResponseListener(response -> {
            alertsShown = true;
            if (response.getAlerts().isEmpty()) {
                // No alerts received, launch app
                Log.i("SplashScreen", "No alerts received, start app");
                onReadyForLogin();
                return;
            }
            // Start alert activity for each alert
            for (Alert alert: response.getAlerts()) {
                Intent alertIntent = new Intent(SplashActivity.this, AlertActivity.class);
                Gson gson = new Gson();
                String alertJson = gson.toJson(alert);
                alertIntent.putExtra("alert", alertJson);
                alertActivityLauncher.launch(alertIntent);
            }
        });
        alertsApiTask.execute(this);
    }

}

File: app/src/main/java/com/allybros/superego/activity/splash/SplashActivity.kt
package com.allybros.superego.activity.splash

import android.content.*
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import com.allybros.superego.R
import android.net.ConnectivityManager
import android.util.Log
import androidx.activity.viewModels
import androidx.appcompat.app.AlertDialog
import androidx.databinding.DataBindingUtil
import com.allybros.superego.api.LoadProfileTask
import com.allybros.superego.unit.ConstantValues
import com.allybros.superego.unit.ErrorCodes
import com.allybros.superego.api.LoginTask
import com.google.android.material.snackbar.Snackbar
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import kotlin.jvm.Synchronized
import com.allybros.superego.activity.userpage.UserPageActivity
import com.allybros.superego.activity.login.LoginActivity
import com.allybros.superego.databinding.ActivitySplashBinding
import com.android.volley.toolbox.Volley
import com.allybros.superego.unit.Trait
import com.allybros.superego.util.SessionManager
import com.android.volley.Request
import com.android.volley.toolbox.StringRequest
import org.json.JSONObject
import org.json.JSONException
import java.util.ArrayList
import kotlin.system.exitProcess

class SplashActivity : AppCompatActivity() {
    private var loadProfileRegister: BroadcastReceiver? = null
    private var loginReceiver: BroadcastReceiver? = null
    var loadTaskLock = false
    var getTraitsLock = false

    lateinit var binding: ActivitySplashBinding
    val viewModel: SplashVM by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_splash)

        // Check internet connection
        val cm = applicationContext.getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
        val activeNetwork = cm.activeNetworkInfo
        val isConnected = activeNetwork != null && activeNetwork.isConnectedOrConnecting
        if (isConnected) {
            getAllTraits(applicationContext)
            setupReceivers()
            SessionManager.getInstance().readInfo(applicationContext)
            if (SessionManager.getInstance().sessionToken.isNotEmpty()) {
                // User signed in before
                LoadProfileTask.loadProfileTask(
                    applicationContext,
                    SessionManager.getInstance().sessionToken, ConstantValues.ACTION_LOAD_PROFILE
                )
            } else {
                returnLoginActivity()
            }
        } else {
            AlertDialog.Builder(this@SplashActivity, R.style.SegoAlertDialog)
                .setTitle("insightof.me")
                .setMessage(R.string.error_no_connection)
                .setPositiveButton(
                    getString(R.string.action_ok)) { _, _ ->
                    finish()
                    exitProcess(0)
                }.show()
        }
    }

    private fun setupReceivers() {
        loadProfileRegister = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                when (intent.getIntExtra("status", ErrorCodes.SYSFAIL)) {
                    ErrorCodes.SUCCESS -> {
                        // Profile loaded successfully, current user must be set on SessionManager
                        loadTaskLock = true
                        notifyTaskComplete()
                    }
                    ErrorCodes.SESSION_EXPIRED -> {
                        // Session is not valid, or expired. Try to login again.
                        val uid = SessionManager.getInstance().userId
                        val password = SessionManager.getInstance().password
                        // Check if user login data present
                        if (uid.isEmpty() || password.isEmpty()) {
                            returnLoginActivity()
                        } else {
                            LoginTask.loginTask(this@SplashActivity, uid, password)
                        }
                    }
                    ErrorCodes.SYSFAIL -> AlertDialog.Builder(
                        this@SplashActivity,
                        R.style.SegoAlertDialog
                    )
                        .setTitle("insightof.me")
                        .setMessage(R.string.error_login_failed)
                        .setPositiveButton(
                            getString(R.string.action_ok)) { _, _ ->
                            SessionManager.getInstance().clearSession(
                                applicationContext
                            )
                            returnLoginActivity()
                        }.show()
                }
            }
        }
        loginReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                when (intent.getIntExtra("status", 0)) {
                    ErrorCodes.SUCCESS ->                         // Login succeeded, start load profile task
                        LoadProfileTask.loadProfileTask(
                            applicationContext,
                            SessionManager.getInstance().sessionToken,
                            ConstantValues.ACTION_LOAD_PROFILE
                        )
                    ErrorCodes.CONNECTION_ERROR -> AlertDialog.Builder(
                        this@SplashActivity,
                        R.style.SegoAlertDialog
                    )
                        .setTitle("insightof.me")
                        .setMessage(R.string.error_no_connection)
                        .setPositiveButton(
                            getString(R.string.action_ok)) { _, _ ->
                            finish()
                            exitProcess(0)
                        }.show()
                    else -> {
                        Snackbar.make(
                            window.decorView,
                            R.string.error_login_again,
                            2000
                        ).show()
                        returnLoginActivity()
                    }
                }
            }
        }
        LocalBroadcastManager.getInstance(applicationContext)
            .registerReceiver(loginReceiver!!, IntentFilter(ConstantValues.ACTION_LOGIN))
        LocalBroadcastManager.getInstance(applicationContext)
            .registerReceiver(loadProfileRegister!!, IntentFilter(ConstantValues.ACTION_LOAD_PROFILE))
    }

    /**
     * Checks if both task are completed. If both of them completed, starts UserPageActivity
     */
    @Synchronized
    private fun notifyTaskComplete() {
        if (loadTaskLock && getTraitsLock) {
            // Both task are completed
            LocalBroadcastManager.getInstance(applicationContext)
                .unregisterReceiver((loadProfileRegister)!!)
            LocalBroadcastManager.getInstance(applicationContext)
                .unregisterReceiver((loginReceiver)!!)
            val i = Intent(this@SplashActivity, UserPageActivity::class.java)
            i.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            startActivity(i)
            finish()
        }
    }

    /**
     * Starts login activity and finished this activity
     */
    private fun returnLoginActivity() {
        LocalBroadcastManager.getInstance(applicationContext)
            .unregisterReceiver((loadProfileRegister)!!)
        LocalBroadcastManager.getInstance(applicationContext).unregisterReceiver((loginReceiver)!!)
        val i = Intent(this@SplashActivity, LoginActivity::class.java)
        i.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        startActivity(i)
        finish()
    }

    /**
     * Sets all trait list
     * @param context    require for sending request
     */
    private fun getAllTraits(context: Context?) {
        val queue = Volley.newRequestQueue(context)
        val traits = ArrayList<Trait>()
        val jsonRequest = StringRequest(
            Request.Method.GET,
            ConstantValues.ALL_TRAITS, { response ->
                Log.d("getAllTraits",response.toString());
                try {
                    val jsonObject = JSONObject(response)
                    for (i in 0 until jsonObject.getJSONArray("o").length()) {
                        val iter = jsonObject.getJSONArray("o")[i] as JSONObject
                        val traitNo: Int = iter.getInt("traitNo")
                        val positiveName: String? = iter.getString("positive")
                        val negativeName: String? = iter.getString("negative")
                        val positiveIcon: String? = iter.getString("positive_icon")
                        val negativeIcon: String? = iter.getString("negative_icon")
                        traits.add(
                            Trait(
                                traitNo,
                                positiveName,
                                negativeName,
                                positiveIcon,
                                negativeIcon
                            )
                        )
                    }
                    for (i in 0 until jsonObject.getJSONArray("c").length()) {
                        val iter = jsonObject.getJSONArray("c")[i] as JSONObject
                        val traitNo: Int = iter.getInt("traitNo")
                        val positiveName: String? = iter.getString("positive")
                        val negativeName: String? = iter.getString("negative")
                        val positiveIcon: String? = iter.getString("positive_icon")
                        val negativeIcon: String? = iter.getString("negative_icon")
                        traits.add(
                            Trait(
                                traitNo,
                                positiveName,
                                negativeName,
                                positiveIcon,
                                negativeIcon
                            )
                        )
                    }
                    for (i in 0 until jsonObject.getJSONArray("e").length()) {
                        val iter = jsonObject.getJSONArray("e")[i] as JSONObject
                        val traitNo: Int = iter.getInt("traitNo")
                        val positiveName: String? = iter.getString("positive")
                        val negativeName: String? = iter.getString("negative")
                        val positiveIcon: String? = iter.getString("positive_icon")
                        val negativeIcon: String? = iter.getString("negative_icon")
                        traits.add(
                            Trait(
                                traitNo,
                                positiveName,
                                negativeName,
                                positiveIcon,
                                negativeIcon
                            )
                        )
                    }
                    for (i in 0 until jsonObject.getJSONArray("a").length()) {
                        val iter = jsonObject.getJSONArray("a")[i] as JSONObject
                        val traitNo: Int = iter.getInt("traitNo")
                        val positiveName: String? = iter.getString("positive")
                        val negativeName: String? = iter.getString("negative")
                        val positiveIcon: String? = iter.getString("positive_icon")
                        val negativeIcon: String? = iter.getString("negative_icon")
                        traits.add(
                            Trait(
                                traitNo,
                                positiveName,
                                negativeName,
                                positiveIcon,
                                negativeIcon
                            )
                        )
                    }
                    for (i in 0 until jsonObject.getJSONArray("n").length()) {
                        val iter = jsonObject.getJSONArray("n")[i] as JSONObject
                        val traitNo: Int = iter.getInt("traitNo")
                        val positiveName: String? = iter.getString("positive")
                        val negativeName: String? = iter.getString("negative")
                        val positiveIcon: String? = iter.getString("positive_icon")
                        val negativeIcon: String? = iter.getString("negative_icon")
                        traits.add(
                            Trait(
                                traitNo,
                                positiveName,
                                negativeName,
                                positiveIcon,
                                negativeIcon
                            )
                        )
                    }
                    getTraitsLock = true
                    notifyTaskComplete()
                    Trait.setAllTraits(traits)
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }
        ) {
            getTraitsLock = true
            notifyTaskComplete()
        }
        queue.add(jsonRequest)
    }
}